pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,389 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storecache
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/thanos-io/thanos/pkg/cache"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type CachingBucketConfig struct {
+	// Basic unit used to cache chunks.
+	ChunkBlockSize int64 `yaml:"chunk_block_size"`
+
+	// Maximum number of GetRange requests issued by this bucket for single GetRange call. Zero or negative value = unlimited.
+	MaxChunksGetRangeRequests int `yaml:"max_chunks_get_range_requests"`
+
+	// TTLs for various cache items.
+	ChunkObjectSizeTTL time.Duration `yaml:"chunk_object_size_ttl"`
+	ChunkBlockTTL      time.Duration `yaml:"chunk_block_ttl"`
+}
+
+func DefaultCachingBucketConfig() CachingBucketConfig {
+	return CachingBucketConfig{
+		ChunkBlockSize:            16000, // Equal to max chunk size.
```

Comment:
Unfortunately this would lead to import cycle. We could move `maxChunkSize` to some other package, I'm not sure it's worth it though.

```
import cycle not allowed
package github.com/thanos-io/thanos/cmd/thanos
	imports github.com/thanos-io/thanos/pkg/query
	imports github.com/thanos-io/thanos/pkg/store
	imports github.com/thanos-io/thanos/pkg/store/cache
	imports github.com/thanos-io/thanos/pkg/store
```

Reply:
I haven't considered the cyclic dependency. The correlation here is quite low with the `maxChunkSize` so I'm OK to keep it as i.